### ISSUE TITLE
Proposal for an advanced decentralized friend system

### ISSUE THREAD

**Issue by Dune-jr** [teeworlds/teeworlds#2423]
## Motivation
At the moment, the friend system is very limited, and we regularly get people asking for a way to chat with friends or just get notified when they're online, like it works in many games. Most of the social hast to happen outside of the game.
To that, we often answer that this would require a centralized server dealing with those things, and this comes with loads of trouble.

So here I drafted some thoughts on how the friend system could be enhanced to

- allow offline chat
- get notifications
  - friend goes online
  - friend invites you to a server
- remain decentralized

## Draft
![sketch12](https://user-images.githubusercontent.com/355114/73034351-88a9a300-3e3c-11ea-82e8-25a2c66fc1cd.png)
(click to maximize)

### Friend requests
Above is a draft of what would happen for a friend request
- A asks the server to send B a friend request
- the server notifies B that A requested friendship, sends A's IP, `@A`
- B manually accepts: B sends to `@A` his IP
- A and B have stored each other's IP, are effectively friends and are free to communicate in many ways

### Tokens
Tokens would be useful (necessary) to identify the friendship. This could be done by saying "I'm "B"", but there are issues with this
  - a malicious "C" could do that, hoping that a friend request to "B" was sent, and give his own IP, effectively spoofing B
  - people can rename themselves, then it gets confusing

### P2P communication

Online statuses could be achieved either by notifying every friend when they are online, by pings/timeouts, or both.

Befriending the same person on different computers (different IP or not) would be possible, different tokens would be generated.

Friend invites would be very straight forward.

This token draft could be improved to a key system, allowing encrypted chat, although I'm not sure that's very useful.

### Backwards compatibility

The server knows the version of the client, and can raise an error when trying to befriend players with older 0.7 versions.

## Thoughts

Those thoughts are not very mature, but I'd love to get some early feedback. Apart from the server nicely forwarding the IP to B, this could be written entirely as a client modification for prototyping, only it would have to be sent as a `!friendreq("A", @A)` whisper message.

**Comment by AssassinTee** [comment]
I see some problems here:
- Server sending fake requests - Guess in my next mod i am friend to everyone
- Server sending fake invites - Friend xy invites you to play gctf
- IPs aren't static and for some players this feature wouldn't make sense

**Comment by Dune-jr** [comment]
> * Server sending fake invites - Friend xy invites you to play gctf

Not possible, game invites are P2P in this draft

> * Server sending fake requests - Guess in my next mod i am friend to everyone

It is true the server can fake a request, and the you would be one-way friend with someone, but you would still have to accept the request, and the friendship wouldn't "work" unless the other person also accepts, with the same token (and then it's fine?).
So all a malicious server can really do is leak the IPs of other players, which it could already be doing now.

> * IPs aren't static and for some players this feature wouldn't make sense

Some players have frequent changing IPs, indeed. For those, we could add a `replace_ip` packet that sends `(old ip, token, new ip)` to all friends.

**Comment by AssassinTee** [comment]
> Not possible, game invites are P2P in this draft

A invites B as friend, but Server sends fake accept (Message never reaches B), so A is befriended with Server.
Server therefore can invite A to play on his server.

![iamahackerman](https://i.imgur.com/PDavaFA.jpg)

> Some players have frequent changing IPs, indeed. For those, we could add a `replace_ip` packet that sends `(old ip, token, new ip)` to all friends.

friend changed ip too, **oops**

**Comment by Dune-jr** [comment]
> > Not possible, game invites are P2P in this draft
> 
> A invites B as friend, but Server sends fake accept (Message never reaches B), so A is befriended with Server.
> Server therefore can invite A to play on his server.
> 
> ![iamahackerman](https://camo.githubusercontent.com/6fefd316d25363953a929ebe2e96dcec8cb30be5/68747470733a2f2f692e696d6775722e636f6d2f504461766146412e6a7067)
> 
> ```
> ACCEPT
> (B, @ServerIP)
> +TOKEN
> ```

That is true, the malicious server can send its own address instead of B (well, some other address, or it would be obvious), and then invite you pretending to be B.
But then you aren't friends with B and the server pretending to be B isn't going to pass the Turing test for long at all, so the nuisance potential seems limited to me. All it would have achieved are a few invites before being identified as malicious, is that worth it?
I mean, if the server wants to be annoying, it can already spam loud sounds or something.

> > Some players have frequent changing IPs, indeed. For those, we could add a `replace_ip` packet that sends `(old ip, token, new ip)` to all friends.
> 
> friend changed ip too, **oops**

Yeah, you're right, if both players change IP, the P2P link is broken.

> On a site note: Server inviting players would be cool, too :3

You don't want to accept invite from non-friends, so that would mean the ability to befriend servers. When would you a want a server to automatically invite you though? That'd probably just be ads.

**Comment by AssassinTee** [comment]
Without a certificate system a [MITM] https://en.wikipedia.org/wiki/Man-in-the-middle_attack is always possible

**Comment by Dune-jr** [comment]
> Without a certificate system a [MITM](https://en.wikipedia.org/wiki/Man-in-the-middle_attack) is always possible

Doesn't this also apply to the current rcon authentification protocol?

**Comment by Robyt3** [comment]
I like the idea of a global friend and chat system, but I think it would be more practical if we already had #146 using some standard authentication protocol (like OAuth2) with accounts managed by teeworlds.com, optionally also reusing existing authentication providers like steam.

**Comment by jxsl13** [comment]
the proposal is about a decentralized system, because Teeworlds tries to be based on decentralization.

**Comment by Dune-jr** [comment]
> I like the idea of a global friend and chat system, but I think it would be more practical if we already had #146 using some standard authentication protocol (like OAuth2) with accounts managed by teeworlds.com, optionally also reusing existing authentication providers like steam.

That's a good point. #146 says it should be implemented with XMPP, which is pretty robust as far as I know.
Indeed it would be more practical and would enable message logs, etc.
I'm not familiar enough with that to figure how much more difficult it would be to implement, as you would also have to run servers for it.

**Comment by Dune-jr** [comment]
> optionally also reusing existing authentication providers like steam.

I'm a bit skeptical about that because that would let servers disallow anyone who is not identified on Steam, and if that catches on with mods, then a large part of Teeworlds would be inaccessible without a Steam account, which is really not in with the spirit of Teeworlds.

**Comment by Robyt3** [comment]
> I'm a bit skeptical about that because that would let servers disallow anyone who is not identified on Steam, and if that catches on with mods, then a large part of Teeworlds would be inaccessible without a Steam account, which is really not in with the spirit of Teeworlds.

Servers should not know any details about client accounts. At the max, there should be a unique ID for a registered account so servers could use this instead of their own registration system, but then there needs to be some generated ID for unregistered players or the servers would force the registration system, which I would also prefer to be optional.

**Comment by Dune-jr** [comment]
The registration server was always meant to be optional, but if I remember correctly it should allow to authenticate people, e.g. something that would let me know that the `/\ssa` player on this server is [teeworlds.com/Assa](https://www.teeworlds.com/forum/profile.php?id=132421).

This would mean that you can't really attribute each player an anonymous ID, the sever will at least know that the player is unregistered.

**Comment by Robyt3** [comment]
Yeah, the servers would always be able to detect unregistered clients, unless this identity confirmation was also P2P. Basically teeworlds.com creates certificates for registered users that they fetch when they login using their teeworlds account. The players can then use the certificates to confirm their identity to each other by signing a message. Once you confirmed the identity of a friend you could also use this for encrypted P2P chat. But having the game servers relay that information would be easier, although you'd have to trust the servers with that data.

**Comment by oy** [comment]
Some concerns:
Like mentioned above some countries/ISP's provide dynamic IP's which change a lot and would be problematic for this.
Sharing IP's with other people and allowing direct player communication would also create new possible ways to attack players.

On a totally different note: in 0.5 times I had a irc client integrated in my tw client for outside communication.
Maybe using some kinda already given infrastructure would be helpfull here?

**Comment by Dune-jr** [comment]
Indeed my biggest concern is that something as innocent as accepting a friend invite, or convincing someone who has you as a friend, would leak your IP to a peer.

> On a totally different note: in 0.5 times I had a irc client integrated in my tw client for outside communication.
Maybe using some kinda already given infrastructure would be helpfull here?

IRC integration would be cool, but we would need to set up some server infrastructure for that. Basic IRC leaks everyone's IPs, so it doesn't solve the problem by itself.

**Comment by oy** [comment]
IRC was just an example, though quakenet lets you hide your hostname.

**Comment by Dune-jr** [comment]
XMPP has its uses in online games for that exact purpose: https://xmpp.org/uses/gaming.html, but I'm not sure that counts as an "already given infrastructure" that we can take advantage of.
